thinc/layers/with_array.py

```diff
@@ -66,24 +68,24 @@ def _get_array(model, X: SeqT) -> ArrayXd:
 
 
 def _list_forward(
-    model: Model[List2d, List2d], Xs: List2d, is_train: bool
-) -> Tuple[List2d, Callable]:
-    layer = model.layers[0]
+    model: Model[SeqT, SeqT], Xs: ListXd, is_train: bool
+) -> Tuple[ListXd, Callable]:
+    layer: Model[ArrayXd, ArrayXd] = model.layers[0]
     pad = model.attrs["pad"]
     lengths = layer.ops.asarray1i([len(seq) for seq in Xs])
-    Xf = layer.ops.flatten(Xs, pad=pad)  # type: ignore
+    Xf = layer.ops.flatten(Xs, pad=pad)
     Yf, get_dXf = layer(Xf, is_train)
 
-    def backprop(dYs: List2d) -> List2d:
-        dYf = layer.ops.flatten(dYs, pad=pad)  # type: ignore
+    def backprop(dYs: ListXd) -> ListXd:
+        dYf = layer.ops.flatten(dYs, pad=pad)
         dXf = get_dXf(dYf)
         return layer.ops.unflatten(dXf, lengths, pad=pad)
 
     return layer.ops.unflatten(Yf, lengths, pad=pad), backprop
 
 
 def _ragged_forward(
-    model: Model[Ragged, Ragged], Xr: Ragged, is_train: bool
+    model: Model[SeqT, SeqT], Xr: Ragged, is_train: bool
```

Comment:
Why did you rewrite these in the private methods? I feel like we've had this conversation before but I can't easily find it back as the PR history has gotten so long ;-) Right now it feels like `forward()` is doing only half the work by casting the output back to `SeqT` - it could also be casting the input model appropriately like it was doing before?

Reply:
Similar comment for `with_array2d` and some other layers where we have this split up by typing. Whatever we end up deciding, we should make sure it's done in the same consistent way throughout the repo.

Reply:
I don't see that casting the model is semantically correct. In reality there is a single model that is capable of handling a variety of types, and the type of that model doesn't change every time it handles a new input that has one of those types.

Reply:
I see what you mean, and that makes sense. Could you double check whether this principle is applied consistently throughout the code base (of this PR) now?